src/test/scala/zio/nio/ChannelSuite.scala

```diff
@@ -1,54 +1,95 @@
-          server  <- AsynchronousServerSocketChannel()
-          _       <- server.bind(address)
-          worker  <- server.accept
-          _       <- worker.readBuffer(sink)
-          _       <- sink.flip
-          _       <- worker.writeBuffer(sink)
-          _       <- worker.close
-          _       <- server.close
-        } yield ()
-
-      def echoClient: IO[Exception, Boolean] =
-        for {
-          address  <- inetAddress
-          src      <- Buffer.byte(3)
-          client   <- AsynchronousSocketChannel()
-          _        <- client.connect(address)
-          sent     <- src.array
-          _        = sent.update(0, 1)
-          _        <- client.writeBuffer(src)
-          _        <- src.flip
-          _        <- client.readBuffer(src)
-          received <- src.array
-          _        <- client.close
-        } yield sent.sameElements(received)
-
-      val testProgram: IO[Exception, Boolean] = for {
-        serverFiber <- echoServer.fork
-        clientFiber <- echoClient.fork
-        _           <- serverFiber.join
-        same        <- clientFiber.join
-      } yield same
-
-      assert(unsafeRun(testProgram))
-    })
+    section(
+      test("read/write") { () =>
+        val inetAddress = InetAddress.localHost
+          .flatMap(iAddr => SocketAddress.inetSocketAddress(iAddr, 13370))
+
+        def echoServer: IO[Exception, Unit] =
+          for {
+            address <- inetAddress
+            sink    <- Buffer.byte(3)
+            server  <- AsynchronousServerSocketChannel()
+            _       <- server.bind(address)
+            worker  <- server.accept
+            _       <- worker.readBuffer(sink)
+            _       <- sink.flip
+            _       <- worker.writeBuffer(sink)
+            _       <- worker.close
+            _       <- server.close
+          } yield ()
+
+        def echoClient: IO[Exception, Boolean] =
+          for {
+            address  <- inetAddress
+            src      <- Buffer.byte(3)
+            client   <- AsynchronousSocketChannel()
+            _        <- client.connect(address)
+            sent     <- src.array
+            _        = sent.update(0, 1)
+            _        <- client.writeBuffer(src)
+            _        <- src.flip
+            _        <- client.readBuffer(src)
+            received <- src.array
+            _        <- client.close
+          } yield sent.sameElements(received)
+
+        val testProgram: IO[Exception, Boolean] = for {
+          serverFiber <- echoServer.fork
+          clientFiber <- echoClient.fork
+          _           <- serverFiber.join
+          same        <- clientFiber.join
+        } yield same
+
+        assert(unsafeRun(testProgram))
+      },
+      test("read should fail when connection close") { () =>
+        val inetAddress = InetAddress.localHost
+          .flatMap(iAddr => SocketAddress.inetSocketAddress(iAddr, 13370))
+
+        def server: IO[Exception, Boolean] = {
```

Comment:
I gave this branch a run and noticed it doesn't close down the listening socket. This means if you execute `test:run` a second time in the same sbt session, you get an address already bound error. I think you need to use bracketing to make sure the cleanup happens. Very roughly:

```scala
        def server: IO[Exception, Boolean] = {
          for {
            address <- inetAddress
            _  <- AsynchronousServerSocketChannel().bracket(_.close.ignore) { server =>
              server.bind(address) *> server.accept.bracket(_.close.ignore) { worker =>
                worker.read(3) *> worker.read(3)
              }
            }
          } yield false
        }.catchSome {
          case ex: java.io.IOException if ex.getMessage == "Connection reset by peer" =>
            ZIO.succeed(true)
        }
```

Reply:
great catch I will update this.